src/lsptoolshost/roslynLanguageServer.ts

```diff
@@ -694,6 +705,71 @@ export class RoslynLanguageServer {
         );
     }
 
+    private registerDocumentOpenForDiagnostics() {
+        // When a file is opened process any build diagnostics that may be shown
+        this._languageClient.addDisposable(
+            vscode.workspace.onDidOpenTextDocument(async (event) => this._onFileOpened(event))
+        );
+    }
+
+    public clearDiagnostics() {
+        this._diagnosticsReportedByBuild.clear();
+    }
+
+    public async setBuildDiagnostics(buildDiagnostics: Array<[vscode.Uri, vscode.Diagnostic[]]>) {
+        this._allBuildDiagnostics = buildDiagnostics;
+        const buildOnlyIds = await this.getBuildOnlyDiagnosticIds(CancellationToken.None);
+        const displayedBuildDiagnostics = new Array<[vscode.Uri, vscode.Diagnostic[]]>();
+
+        this._allBuildDiagnostics.forEach((fileDiagnostics) => {
+            const uri = fileDiagnostics[0];
+            const diagnosticList = fileDiagnostics[1];
+
+            // Check if we have live diagnostics shown for this document
+            const liveDiagnostics = vscode.languages.getDiagnostics(uri);
```

Comment:
I'm not sure if this check is correct - it could be the case that we have no live diagnostics but we should show only build diagnostics (e.g. the errors were already fixed during the build)

Perhaps it would be better to check the state of the FSA flag - if FSA is workspace, then we know live diagnostics should win.

And of course open document diagnostics should always come from live.

Reply:
Though this has the risk that build diagnostics don't get shown for a while (if workspace diagnostics are slow to return)

Reply:
Hmm the de-dupe logic in the non-LSP implementation is slightly different. Once diagnostics from build are intercepted, we report the ones with build-only diagnostic ID via a build-only error list source. For the rest of them, we throw away all the currently computed/cached live diagnostics (both for open and closed documents) and replace them with the non-build only ones that came from build. These are instantly refreshed on the error list, so your error list content will exactly match the build output window content when a build finishes. Subsequently, we trigger live re-analysis of active and open documents, so in case anything has changed in the snapshot after the build was kicked, we should get refreshed diagnostics for these changed files. Closed files continue having the diagnostics from build, until they get refreshed from FSA analysis (if enabled) OR the document is opened.

The approach chosen here seems to be also fine - though we need to now handle any timing related issues, FSA state, etc. that @dibarbet mentioned

Reply:
Tracking this here https://github.com/dotnet/vscode-csharp/issues/6606